user_status: Allow users to set time to clear their status.

Vector73 opened the pull request:
This commit adds the feature wherein users could select a time at which their status would be cleared, while setting the status.

Fixes #28543

<!-- Describe your pull request here.-->
<details><summary>Screenshots</summary>
<p>

<img width="530" height="663" alt="image" src="https://github.com/user-attachments/assets/76e5fdfb-2810-46b8-8d00-a8ff593bb5d9" />

<img width="520" height="629" alt="image" src="https://github.com/user-attachments/assets/f5c3a854-132b-4157-90f4-b7f1d5590e74" />

</p>
</details>

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! We need more vertical space above the new setting (whatever is standard in our modals).

alya commented:
Let's use the same styling as on the invitation modal:

<img width="433" height="88" alt="Screenshot 2025-10-23 at 10 48 33" src="https://github.com/user-attachments/assets/5becd622-84e6-4cc0-9396-ce87a8b3f3cb" />

I think we can try leaving just the time with no extra words before it, like we have now.

alya commented:
The dropdown option labels should respect the 24 hr time preference:

<img width="247" height="91" alt="Screenshot 2025-10-23 at 10 52 58" src="https://github.com/user-attachments/assets/b582a845-61a4-498e-939a-e71a0b181de2" />

alya commented:
If you set your status to clear in the past, it resets to right now in the UI (which is probably fine), but then it doesn't clear. When I open the modal again, it's set to "Never", which is not expected. Probably if the configured time is in the past, we should just remove it immediately.

sahil839 commented on web/src/user_status.ts:
```diff
@@ -115,6 +160,7 @@ export function initialize(params: StateData["user_status"]): void {
         if (dct.status_text) {
             user_info.set(user_id, dct.status_text);
         }
+        set_scheduled_end_time(dct.scheduled_end_time);
```
This results in a bug where `scheduled_end_time` is set to `undefined` if there is some other user than the user themself who has status set and the user themself is not the last user called in this loop.

Vector73 replied:
Fixed it now.

sahil839 reviewed (commented):
Posted some comments.

sahil839 commented on zerver/tests/test_user_status.py:
```diff
@@ -196,6 +204,96 @@ def update_status_and_assert_event(
+            emoji_code="1f697",
+            reaction_type=UserStatus.UNICODE_EMOJI,
+            client_id=client1.id,
+            scheduled_end_time=scheduled_timestamp_expired,
```
I think this can also be replaced by calling the API endpoint.

sahil839 commented on zerver/lib/user_status.py:
```diff
@@ -45,6 +53,8 @@ def format_user_status(row: RawUserInfoDict) -> UserInfoDict:
         dct["emoji_name"] = emoji_name
         dct["emoji_code"] = emoji_code
         dct["reaction_type"] = reaction_type
+    if row["user_profile_id"] == user_id and scheduled_end_time:
```
Does this mean that we will not include `scheduled_end_time` field if the status is set to never expire? Was this discussed?

Vector73 replied:
Yeah, if the `scheduled_end_time` is set to "no end time", it will not be included, but it is not discussed. I think it is consistent with how we handle other fields -- not included in the dictionary if not set.

sahil839 replied:
We pass `expiry_time` field for invites even when invite is set to never expire. I think it is good to discuss such API changes once to avoid changes later which require handling backwards compatibility.

sahil839 commented on web/src/user_status_ui.ts:
```diff
@@ -48,6 +132,28 @@ export function open_user_status_modal(): void {
+                },
+            ),
+        );
+        $("#scheduled-end-time-description").toggleClass("invisible", true);
```
We mostly use `.show`/`.hide` or toggle `hide` class for showing and hiding elements. Can we do that here as well?

sahil839 commented on web/src/user_status_ui.ts:
```diff
@@ -185,35 +303,90 @@ function user_status_post_render(): void {
+        const now = new Date();
+        now.setSeconds(0, 0);
+        flatpickr.show_flatpickr(
+            $(".dropdown-list-item-common-styles")[4]!,
```
Looks weird to select the element for flatpickr like this by using index in array of elements. Can we use the dropdown widget button as the element itself here?

Vector73 commented on zerver/lib/user_status.py:
```diff
@@ -125,10 +142,42 @@ def get_user_status(user_profile: UserProfile) -> UserInfoDict:
+        and ((old_status_text in ["", None] and status_text is None) or status_text == "")
+    ) and ((old_emoji_name in ["", None] and emoji_name is None) or emoji_name == ""):
+        raise JsonableError(_("Client does not have any status set."))
+
```
OK, I have moved it to `views/user_status.py`.

sahil839 commented on web/src/server_events_dispatch.js:
```diff
@@ -1198,6 +1195,9 @@ export function dispatch_normal_event(event) {
                 );
             }
+            if (event.user_id === current_user.user_id) {
+                user_status.set_scheduled_end_time(event.scheduled_end_time);
```
I think this will set `scheduled_end_time` to undefined if it is not passed in the event when it is not updated. So, would be better to have a check for undefined value as well.

sahil839 reviewed (commented):
Have not reviewed frontend code again but just posted some comments about refactoring.

Vector73 commented:
Addressed the recent feedback.

sahil839 commented on zerver/views/presence.py:
```diff
@@ -123,6 +130,29 @@ def update_user_status_backend(
+            emoji_name = ""
+            scheduled_end_time = None
+
+        elif ((old_status_text in ["", None] and status_text is None) or status_text == "") and (
```
> This logic is a bit flawed. It will not raise error for some cases, for example, `current_status={"status_text": "abc"}` `status_text=""` `emoji_name=None`
> 
> We need to individually check for both `status_text` and `emoji_name` being empty before throwing the error.

I don't get this fully. Are the only fields passed to API request are the one you mentioned above or there are some more? Assuming no other fields in request, initially `status_text` is set to "abc", and then user tries to remove the status by setting `status_text` to "".

sahil839 replied:
Ohk. It still looks hard to understand. Will this modification work -
```py
    current_status = get_user_status(user_profile)
    if not isinstance(scheduled_end_time, MissingType):
        if scheduled_end_time is not None and scheduled_end_time <= datetime_to_timestamp(timezone_now()):
            # Clear out the status in case status is set with an already
            # expired time.
            status_text = ""
            emoji_name = ""
            scheduled_end_time = None

        if scheduled_end_time is not None:
            if current_status == {}:
                if status_text in ["", None] and emoji_name in ["", None]:
                    raise JsonableError(_("Client must set status if they pass scheduled_end_time."))

            else:
                proposed_emoji_name = emoji_name if emoji_name is not None else current_status["emoji_name"]
                proposed_status_text = status_text if status_text is not None else current_status["status_text"]
                if proposed_status_text == "" and proposed_emoji_name == "":
                    raise JsonableError(_("Client must set status if they pass scheduled_end_time.")) 

        if current_status != {} and scheduled_end_time == current_status.get("scheduled_end_time"):
            # Set value as Missing if scheduled_end_time is same as before.
            scheduled_end_time =  Missing

    old_status_text = current_status.get("status_text")
    old_emoji_name = current_status.get("emoji_name")
```

Vector73 replied:
Yeah, I guess that could work. I have further simplified it -- would be great if you could take a look again.

sahil839 commented on web/src/user_status.ts:
```diff
@@ -21,16 +21,29 @@ const user_status_event_schema = z.intersection(
 const user_info = new Map<number, string>();
 const user_status_emoji_info = new Map<number, UserStatusEmojiInfo>();
 
+let scheduled_end_time: number | undefined | null;
```
Can we avoid setting this to undefined?
I think this will be undefined when user has not set any time to clear the status or has not set the status yet . Can we just keep the value for this as `null` in both the cases?

Then the code below to call `set_scheduled_end_time` can also have the condition to check if `dct.scheduled_end_time !== undefined`.

Vector73 replied:
OK, I have updated the code.

sahil839 commented on web/src/user_status_ui.ts:
```diff
@@ -36,6 +115,11 @@ export function open_user_status_modal(): void {
 
+    scheduled_end_time = user_status.get_scheduled_end_time();
+    if (scheduled_end_time) {
+        scheduled_end_time *= 1000;
```
A comment might be helpful here.

Vector73 replied:
Added `// Convert scheduled_end_time to milliseconds.`

timabbott replied:
I think it might make more sense for `user_status.get_scheduled_end_time` to return a time in the units that calling code expects to use it in? I'm also a bit confused how this is in milliseconds in first place, since I'd expect it to have been a date, not an offset.

sahil839 reviewed (commented):
Posted a couple of minor comments. Looks good otherwise.

alya commented:
Hm, why am I getting an error when trying to test (I did provision)?

```
ProgrammingError at /
column zerver_userstatus.scheduled_end_time does not exist
LINE 1: ...r_userstatus"."reaction_type" AS "reaction_type", "zerver_us...


Exception Location: | /srv/zulip/zerver/lib/db.py, line 22, in wrapper_execute
-- | --
zerver.views.home.home

```

alya commented:
Cool, yeah, `--force` fixed it.

alya commented:
Not a blocker for merging the main PR, but we should probably show the currently selected option in bold/highlighted in the dropdown menu, like we do, e.g., here.

<img width="448" height="248" alt="Screenshot 2025-11-13 at 13 38 05@2x" src="https://github.com/user-attachments/assets/a68e891e-abcb-48cc-89b6-9c089d051705" />

timabbott commented on tools/run-dev:
```diff
@@ -204,6 +204,7 @@ def server_processes() -> list[list[str]]:
             "--quiet",
         ],
         ["./manage.py", "deliver_scheduled_messages", "--automated"],
+        ["./manage.py", "clear_user_status"],
```
Hmm, is there a way we can avoid adding another process in the development environment here? We're trying to keep resources minimal in the development environment.

@alexmv FYI in case you have ideas for what we should do. Options I see include:
- Make `manage.py process_queue` incorporate these non-queue workers.
- Merging this with `deliver_scheduled_messages` at least.

(There's kinda a similar goal of avoiding an extra process for low-RAM self-hosted systems)

timabbott commented on web/src/user_status_ui.ts:
```diff
@@ -1,19 +1,96 @@
+        case "today_five_pm":
+            return today.setHours(17, 0, 0, 0);
+        case "tomorrow":
+            return tomorrow.setHours(0, 0, 0, 0);
```
Have we discussed the decision to have this correspond to midnight in #design?  That seems like a thing that should be considered; a "tomorrow morning" that changes at like 5AM might be more popular.

timabbott commented on web/src/user_status_ui.ts:
```diff
@@ -48,6 +131,28 @@ export function open_user_status_modal(): void {
+    if (scheduled_end_time) {
+        $("#scheduled_end_time_widget .dropdown_widget_value").text(
+            $t(
+                {defaultMessage: "{N}"},
```
I don't understand what the translators are being asked to translate here; is it the curly braces?

timabbott commented on zerver/lib/user_status.py:
```diff
@@ -22,16 +26,18 @@ class RawUserInfoDict(TypedDict):
 
 
-def format_user_status(row: RawUserInfoDict) -> UserInfoDict:
+def format_user_status(user_id: int, row: RawUserInfoDict) -> UserInfoDict:
```
Maybe call this `acting_user_id`, since that's terminology we use to indicate that the user ID specifically refers to the current acting user.

timabbott commented on zerver/lib/user_status.py:
```diff
@@ -45,6 +51,8 @@ def format_user_status(row: RawUserInfoDict) -> UserInfoDict:
         dct["emoji_code"] = emoji_code
         dct["reaction_type"] = reaction_type
+    if row["user_profile_id"] == user_id and scheduled_end_time:
+        dct["scheduled_end_time"] = datetime_to_timestamp(scheduled_end_time)
```
I feel like this block wants a comment explaining the conditionals. Something like:

> Scheduled end times are not shared with other users.

timabbott commented:
Reviewed. I think everything but https://github.com/zulip/zulip/pull/36352#discussion_r2525263305 should be pretty easy to resolve. We may want a dedicated #backend conversation to decide what to do with that; feel free to start one and cross-link.

andersk commented on zerver/actions/user_status.py:
```diff
@@ -35,21 +40,58 @@ def do_update_user_status(
+    )
+
+    if user_statuses.exists():
+        for user_status in user_statuses:
```
`.exists()` wastefully runs a separate database query. (It would be more efficient if we were not also loading the data, but we are.) Do this instead.

```suggestion
    if user_statuses:
        for user_status in user_statuses:
```

Vector73 replied:
I guess the `if` condition can be removed altogether, as there is no return value for this function now.

andersk commented on web/src/user_status.ts:
```diff
@@ -21,16 +21,35 @@ const user_status_event_schema = z.intersection(
     emoji_name: string;
     emoji_code: string;
     reaction_type?: string;
+    scheduled_end_time: number | string;
```
This should take `number | null`, and should take responsibility for `JSON.stringify`ing it here, not in the caller.

andersk commented on web/src/user_status_ui.ts:
```diff
@@ -1,19 +1,96 @@
+        case "in_thirty_minutes":
+            return new Date(now).setMinutes(now.getMinutes() + 30);
+        case "in_one_hour":
+            return new Date(now).setHours(now.getHours() + 1);
```
This gives incorrect results near daylight saving time transitions, e.g., in `America/Los_Angeles`, `2025-11-02T08:30Z` → `2025-11-02T10:30Z`.

Vector73 replied:
Oh, fixed it now. BTW, this is the first time I heard about daylight saving time. I read about it, sounded like a really weird concept :)

Vector73 commented on zerver/management/commands/clear_user_status.py:
```diff
@@ -0,0 +1,40 @@
+        try:
+            while True:
+                if try_clear_scheduled_user_status():
+                    continue
```
OK, removed the return value.

Vector73 committed:
user_status: Allow users to set time to clear their status.

This commit adds the feature wherein users could select a time
at which their status would be cleared, while setting the status.

Fixes #28543

Co-authored-by: roanster007 <rohan.gudimetla07@gmail.com>

Asteriskkkk commented:
@Vector73 are you up for resolving the conflicts or should i take it over. i think this is quite close to getting merged.

peter-emtelligent commented:
Any chance this could be merged any time soon? Would be nice to have this functionality!

Asteriskkkk commented:
@peter-emtelligent  there hasn't been any discussion on this, opening a thread in [CZO](https://chat.zulip.org/) would be good to get attention on this.
